clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2949,15 +2950,34 @@ static void handleSectionAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   }
 }
 
+static bool isValidCodeModelAttr(Sema &S, StringRef Str) {
+  if (S.Context.getTargetInfo().getTriple().isLoongArch()) {
+    return Str == "normal" || Str == "medium" || Str == "extreme";
+  } else {
+    assert(S.Context.getTargetInfo().getTriple().getArch() ==
+               llvm::Triple::x86_64 &&
+           "only loongarch/x86-64 supported");
+    return Str == "small" || Str == "large";
+  }
+}
+
 static void handleCodeModelAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   StringRef Str;
   SourceLocation LiteralLoc;
   // Check that it is a string.
   if (!S.checkStringLiteralArgumentAttr(AL, 0, Str, &LiteralLoc))
     return;
 
+  // Ignore the attribute for GPU device compiles since it only applies to host
+  // globals.
+  if (S.Context.getTargetInfo().getTriple().isNVPTX() ||
```

Comment:
We should diagnose if this doesn't apply to host and device (see getAuxTriple).  The only valid reason to ignore this is if we see it is not in host, but IS in device (aux).

Reply:
we'll get a warning on the host side in the case the host target doesn't support the attribute and there's no reason to emit the warning again on the device side right?

Reply:
Perhaps?  Perhaps not.  WE aren't necessarily the compiler on both sides, so we're better off making sure we diagnose it as soon as we know what is going on.

Reply:
Aren't we? I mean, I'm sure nvidia and vendors everywhere do crazy stuff, but I would like to clarify what the support scope of the Clang project is. My understanding of all the CUDA functionality we've ever added to Clang was that it assumes we're using the Clang compiler on both sides, not the janky nvcc split pre-processing model.

Reply:
Its not just NVIDIA.  Intel does it, and IIRC, the OMP offload supports arbitrary host compilers as well.  Either way, we should be diagnosing "not supported" as early as we know anyway, which means during 'device' compilation (if it is first).

Reply:
now we warn if both the current target and `getAuxTriple()` don't support the attribute. however, if we're a pure GPU compile with no `getAuxTriple()`, we silently ignore the attribute since we don't know what the host could be. does this work?